### PR TITLE
Implemented the first base for reading codes.

### PR DESCRIPTION
**Pull request by zommiommy**
Defined two traits:
* `WordReader`: Trait of objects that can read `u64` words **in native endianess**
* `ReadBits`: Trait of objects that can read fixed length and unary codes

And an implementation for each:
* `MemWordReader`: Implementer of `WordReader` for `&[u64]`
* `BufferedBitStreamReader`: Implementer of `ReadBits` over a generic type that implements `WordReader`

### CODE REVIEW

**Comment by zommiommy** [comment]
as @zacchiro pointed out, this branch has evolved quite a bit. This could be a good moment to review and merge it as it's in a consistent state. All test pass and all features are fully implemented.

**Comment by zacchiro** [comment]
> as @zacchiro pointed out, this branch has evolved quite a bit. This could be a good moment to review and merge it as it's in a consistent state. All test pass and all features are fully implemented.

+1

@vigna can you please give it a go at reviewing this PR and, if it's OK, merging it? (Ideally preserving commit history, rather than squashing it all together in a single commit.) If not, shout, and let's iterate a bit on improving it before opening a separate one for additional features.

**Comment by zacchiro** [comment]
> I would wait for zeta codes to be implemented so we have everything for webgraph.

It's a clear and verifiable goal, so sounds great to me! Thanks for the feedback.

**Comment by vigna** [comment]
> On 11 Apr 2023, at 20:29, Stefano Zacchiroli ***@***.***> wrote:
> 
> 
> I would wait for zeta codes to be implemented so we have everything for webgraph.
> It's a clear and verifiable goal, so sounds great to me! Thanks for the feedback

(which implies implementing minimal binary codes--Tommaso, can we have a quick chat tomorrow about that?)

In any case, progress has been impressive IMHO.

Ciao,

                                                         seba

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+use std::path::PathBuf;
+
+fn main() {
+    let path: PathBuf = ["src", "codes", "unary_tabls.rs"].iter().collect();
```
Comment (vigna) [inline build.rs:4]: Isn't this a typo?

Comment (zommiommy) [reply build.rs]: yes!

**Pull request merged**